### PR TITLE
Feat: add Higgs Audio TTS

### PR DESCRIPTION
**Pull request by weedge**
audio tokenizer (audio codec):

- https://huggingface.co/bosonai/higgs-audio-v2-tokenizer
- https://github.com/boson-ai/higgs-audio/blob/main/tech_blogs/TOKENIZER_BLOG.md

image

0. waveform -> librosa.load -> pyloudnorm.normalize.loudness -> audio_tokenizer_feature_extractor(HiggsAudioFeatureExtractor) audio features input 
1. Acoustic Encoder(dac2)  + ( Semantic Teacher(bert) -> Semantic Encoder(Encoder) )
2. prior FullConnector(Dense): Linear
3. Quantizer: ResidualVectorQuantizer / ResidualFSQ
4. post FullConnector(Dense): Acoustic Linear , Semantic Linear
5. AcousticDecoder(dac2) , Semantic Decoder(Decoder)
6. Semantic Loss (mse_loss) with (input output audio semantic features)

---
higgs-audio-v2：
- https://huggingface.co/bosonai/higgs-audio-v2-generation-3B-base
- https://github.com/boson-ai/higgs-audio/blob/main/tech_blogs/ARCHITECTURE_BLOG.md
- https://www.boson.ai/blog/higgs-audio-v2

image

0. prepare inputs:
- text encode: text -> text tokenizer encode -> token ids (text/label)
- ref audio encode (one shot): audio(waveform) -> audio tokenizer encode (0,1,2,3) -> vq code (audio_ids)
- audio processor (collator for input to inference and input/output(label) to training/multi-turn-chat) 
  - for Audio-Understanding: audio(waveform) ->  whisper Encoder(feature_extractor) -> audio features/mask
  - vq code (audio_ids) -> build_delay_pattern_mask like MusicGen delay pattern -> new vq code (audio_ids)
  - handle padding for text input ids and attention mask for LLM inference or forward
1. reset kv cache for generate(inference), use StaticCache for torch compile (use [cuda graphs](https://developer.nvidia.com/blog/cuda-graphs/))
2.  LLM forword/generate(inference):
- Extract the text input_ids embeddings: input_ids(text token_id) -> Embedding(embed_tokens) -> embeddings
- Extract audio features embeddings(for Audio-Understanding): audio features -> Whisper Encoder -> last_hidden_state -> HiggsAudioFeatureProjector(Linear) -> audio features embeddings
- Embed the audio(input/output) ids(vq code): audio_ids(audio_codebook vq codes) -> Embedding(audio_codebook_embeddings) -> audio(input/output) embeddings
- Merge text, audio-in embeddings, and audio-out embeddings (merge_input_ids_with_audio_features)
- if use kvcache and use StaticCache, Use the captured cuda graph runner for decoding to run LLM forword (use [torch.cuda.CUDAGraph](https://docs.pytorch.org/docs/stable/generated/torch.cuda.CUDAGraph.html))
- decoding(LLM forword_core): 
  - merged embedding hidden_states + position_ids -> rotary_emb (LlamaRotaryEmbedding) -> position_embeddings
  - ⭐️ merged embedding hidden_states + position_embeddings -> dual_ffn_fast_forward adaptor(HiggsAudioDualFFNDecoderLayer)(28 layers) -> hidden_states
- hidden_states -> audio_decoder_proj(HiggsAudioDecoderProjector) (text Linear + audio Linear) head -> text/audio logits
- if do_sample, logits -> ⭐️ sampling with RAS (_sample) -> text token_id + audio vq code; if use streamer, put generated text token_id and audio vq code to queue
3. text token_id + audio vq code decode: 
- text token_ids decode:  text token_ids -> text tokenizer decode -> text tokens
- audio codebook vq codes decode: audio vq codes -> audio tokenizer decode -> waveform (numpy array)

---
AI podcasts:
- Higgs Audio V2 : https://podcast-997.pages.dev/podcast/c80215a4c3864db58788bd2513d4e606
- EmergentTTS-Eval: https://podcast-997.pages.dev/podcast/4a6f5515d7434b8e86676d59f22db077
- MusicGen: https://podcast-997.pages.dev/podcast/a5cbf999508a4abaa66b57c7d2f5bf93 (gen stream codebook codes style: **delay** pattern， 折中 `flattening` （质量）和 `parallel`（速度）)
- RAS( repetition aware sampling) from VALL-E 2: https://podcast-997.pages.dev/podcast/2bdf1d8c7a97434f8d0f623b5f280ba4
---
> [!TIP]
> - 基于超过 1000 万小时的音频数据和丰富的文本数据进行继续预训练(PT，底座做好充足的准备), 无后训练阶段；（后续提升应该会引入RL， fishaudio S1已经验证了可行）
> - 开源的PT模型，整体感觉对于中文的音频生成质量比英文差些
> - 支持生成BGM和哼唱，实验阶段，不稳定，有时生成会生成很长的音频，直到最大生成token数
> - 语音理解方面，继承自Higgs Audio能力，暂无样例，可以看 https://www.boson.ai/blog/higgs-audio 中的Reasoning(CoT)部分case. (需要修改下[config.json](https://huggingface.co/bosonai/higgs-audio-v2-generation-3B-base/blob/main/config.json)的默认赋值, 需要使用 whisper processor(最新的large-v3-turbo Encoder) 获取 audio features 和 audio features embeddings)
---

feat: 
- add generate stream with delta
- add tts_higgs
- add higgs audio tts test cases on modal
```shell
    Higgs-Audio is an end-to-end multimodal model with the capability to understand and generate text / audio.

    Consider the following example for mixed text/audio understanding / generation:

    - input_tokens: <text_token1><|audio_bos|>[AUDIO]<|audio_eos|><text_token2><|audio_bos|>[AUDIO]<|audio_eos|><text_token4>
    - input_tokens: <text_token1><|audio_bos|>[AUDIO]<|audio_eos|><text_token2><|audio_out_bos|>[AUDIO_OUT]<|audio_eos|><text_token4>

    We will fill [AUDIO] with the audio features extracted by Whisper and fill [AUDIO_OUT] with the audio tokens. when use audio understand

    Consider the following example for mixed text/audio generation:

    text: <|audio_out_bos|>    MASK           MASK           MASK          MASK               MASK         <|audio_eos|> [text_token1]
    audio:     MASK    <|audio_stream_bos|> [audio_token1] [audio_token2] [audio_token3] <|audio_stream_eos|>   MASK           MASK
    token_type: 0               1              1              1             1                  1                 0              0

NOTE:  if use do_sample is True, see _sample code
HiggsAudioModel
    # Built on top of GenerationMixin._sample.
    # We revise the implementation to support generating both audio / text.
    def _sample()
```

```shell
TIP:
- 中文简单的词，比如 "好", 有时没有生成声音（人类所能分辨的音频段)

# put assets
modal volume put assets assets/basic_ref_zh.wav /

# download
modal run src/download_models.py --repo-ids "bosonai/higgs-audio-v2-tokenizer"
modal run src/download_models.py --repo-ids "bosonai/higgs-audio-v2-generation-3B-base"

IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task dump_model

IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task audio_generate
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task audio_generate --example voice_clone

IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task audio_generate_stream

APP_NAME=achatbot IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task achatbot_audio_generate_stream 

IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task run_hf_example --example zero_shot
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task run_hf_example --example voice_clone 
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task run_hf_example --example interleaved_dialogue

# https://github.com/boson-ai/higgs-audio/blob/main/examples/README.md

# ref clone voice
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/single_speaker/en_dl.txt \
    --ref-audio broom_salesman \
    --seed 12345 \
    --out-path ref_broom_salesman.wav
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/single_speaker/en_dl.txt \
    --ref-audio belinda \
    --seed 12345 \
    --out-path ref_belinda.wav
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/single_speaker/en_dl.txt \
    --scene-prompt empty \
    --ref-audio zh_man_sichuan \
    --temperature 0.3 \
    --seed 12345 \
    --out-path ref_zh_man_sichuan.wav

# Random voice
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/single_speaker/en_dl.txt \
    --seed 12345 \
    --out-path random_en.wav
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/single_speaker/zh_ai.txt \
    --seed 12345 \
    --out-path random_zh.wav

# Describe speaker characteristics with text
## Male British Accent
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/single_speaker/en_dl.txt \
    --ref-audio profile:male_en_british \
    --seed 12345 \
    --out-path character_male_en_british.wav
## Female British Accent
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/single_speaker/en_dl.txt \
    --ref-audio profile:female_en_british \
    --seed 12345 \
    --out-path character_female_en_british.wav

# Chunking for long-form audio generation
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --scene-prompt scene_prompts/reading_blog.txt \
    --transcript transcript/single_speaker/en_higgs_audio_blog.md \
    --ref-audio en_man \
    --chunk-method word \
    --temperature 0.3 \
    --generation-chunk-buffer-size 2 \
    --seed 12345 \
    --out-path long_chunk_buffer_audio_blog.wav

# Experimental and Emergent Capabilities
## Hum a tune with the cloned voice
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/single_speaker/experimental/en_humming.txt \
    --ref-audio en_woman \
    --temperature 1.0 \
    --ras-win-len 0 \
    --seed 12345 \
    --out-path en_woman_tune.wav
## Read the sentence while adding background music (BGM)
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/single_speaker/experimental/en_bgm.txt \
    --ref-audio en_woman \
    --temperature 1.0 \
    --ras-win-len 0 \
    --ref-audio-in-system-message \
    --seed 123456 \
    --out-path en_woman_bgm.wav

# Multi-speaker Audio Generation
## Multi-voice random
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/multi_speaker/en_argument.txt \
    --seed 12345 \
    --out-path multi_speaker_random.wav

## Multi-voice clone
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/multi_speaker/en_argument.txt \
    --ref-audio belinda,broom_salesman \
    --ref-audio-in-system-message \
    --chunk-method speaker \
    --seed 12345 \
    --out-path multi_speaker_belinda,broom_salesman.wav
IMAGE_GPU=L4 modal run src/llm/transformers/higgs_audio.py --task generation \
    --transcript transcript/multi_speaker/en_higgs.txt \
    --ref-audio broom_salesman,belinda\
    --ref-audio-in-system-message \
    --chunk-method speaker \
    --chunk-max-num-turns 2 \
    --seed 12345 \
    --out-path multi_speaker_broom_salesman,belinda.wav
```
- add higgs tts grpc client/server
```shell
# put assets
modal volume put assets assets/basic_ref_zh.wav /

# download
modal run src/download_models.py --repo-ids "bosonai/higgs-audio-v2-tokenizer"
modal run src/download_models.py --repo-ids "bosonai/higgs-audio-v2-generation-3B-base"

# run higgs tts server
TTS_TAG=tts_higgs IMAGE_GPU=L4 modal run src/grpc_tts_serve.py

# run grpc client
TTS_TAG=tts_higgs IS_SAVE=1 SERVE_ADDR=r447.modal.host:35591 \
    TTS_AUDIO_TOKENIZER_PATH=/root/.achatbot/models/bosonai/higgs-audio-v2-tokenizer \
    TTS_LM_MODEL_PATH=/root/.achatbot/models/bosonai/higgs-audio-v2-generation-3B-base \
    TTS_REF_TEXT="对，这就是我，万人敬仰的太乙真人。" \
    TTS_REF_AUDIO_PATH="/root/.achatbot/assets/basic_ref_zh.wav" \
    TTS_CHUNK_SIZE=16 \
    python -m src.cmd.grpc.speaker.client

# tips: u can pip install achatbot to run grpc client
ACHATBOT_PKG=1 TTS_TAG=tts_higgs IS_SAVE=1 SERVE_ADDR=r447.modal.host:35591 \
    TTS_AUDIO_TOKENIZER_PATH=/root/.achatbot/models/bosonai/higgs-audio-v2-tokenizer \
    TTS_LM_MODEL_PATH=/root/.achatbot/models/bosonai/higgs-audio-v2-generation-3B-base \
    TTS_REF_TEXT="对，这就是我，万人敬仰的太乙真人。" \
    TTS_REF_AUDIO_PATH="/root/.achatbot/assets/basic_ref_zh.wav" \
    TTS_CHUNK_SIZE=16 \
    python -m achatbot.cmd.grpc.speaker.client
```
---
# reference
- https://github.com/boson-ai/higgs-audio
- https://github.com/boson-ai/emergenttts-eval-public

### CODE REVIEW

**Comment by weedge** [comment]
higgs-audio-v2-generation-3B-base 5771.283456 M parameters

```
HiggsAudioModel(
  (embed_tokens): Embedding(128256, 3072, padding_idx=128001)
  (layers): ModuleList(
    (0-27): 28 x HiggsAudioDualFFNDecoderLayer(
      (self_attn): LlamaSdpaAttention(
        (q_proj): Linear(in_features=3072, out_features=3072, bias=False)
        (k_proj): Linear(in_features=3072, out_features=1024, bias=False)
        (v_proj): Linear(in_features=3072, out_features=1024, bias=False)
        (o_proj): Linear(in_features=3072, out_features=3072, bias=False)
        (rotary_emb): LlamaRotaryEmbedding()
      )
      (mlp): LlamaMLP(
        (gate_proj): Linear(in_features=3072, out_features=8192, bias=False)
        (up_proj): Linear(in_features=3072, out_features=8192, bias=False)
        (down_proj): Linear(in_features=8192, out_features=3072, bias=False)
        (act_fn): SiLU()
      )
      (audio_mlp): LlamaMLP(
        (gate_proj): Linear(in_features=3072, out_features=8192, bias=False)
        (up_proj): Linear(in_features=3072, out_features=8192, bias=False)
        (down_proj): Linear(in_features=8192, out_features=3072, bias=False)
        (act_fn): SiLU()
      )
      (audio_input_layernorm): LlamaRMSNorm((3072,), eps=1e-05)
      (audio_post_attention_layernorm): LlamaRMSNorm((3072,), eps=1e-05)
      (input_layernorm): LlamaRMSNorm((3072,), eps=1e-05)
      (post_attention_layernorm): LlamaRMSNorm((3072,), eps=1e-05)
    )
  )
  (norm): LlamaRMSNorm((3072,), eps=1e-05)
  (rotary_emb): LlamaRotaryEmbedding()
  (audio_decoder_proj): HiggsAudioDecoderProjector(
    (text_lm_head): Linear(in_features=3072, out_features=128256, bias=False)
    (audio_lm_head): Linear(in_features=3072, out_features=8208, bias=False)
  )
  (audio_codebook_embeddings): Embedding(8208, 3072)
)
```
